matchmaking/matchmaker.py

```diff
@@ -0,0 +1,344 @@
+
+        possible_matches = {}
+        for profile_A, topics in match_requests_profile_to_topic.items():
+            possible_matches[profile_A] = set()
+            for topic in topics:
+                profiles_B = match_requests_topic_to_profile[topic]
+                for profile_B in profiles_B:
+                    if profile_A != profile_B:
+                        possible_matches[profile_A].add(profile_B)
+
+        ######################
+        # assign instance vars
+        self._most_recent_match_by_profile = most_recent_match_by_profile
+        # example: {
+        #   ('matt.cronin@gmail.com'): {
+        #       'num_attending': 2,
+        #       'date': datetime.datetime(2020, 10, 3, 0, 0, tzinfo=timezone.utc),
+        #   },  ...
+
+        self._recent_match_by_profile_pair = recent_match_by_profile_pair
+        # example: {
+        #   ('matt.cronin@gmail.com', 'nikola.novakovic9@gmail.com'): { <same as above> } ...
+
+        self._recent_match_by_profile_topic = recent_match_by_profile_topic
+        # example: {
+        #   ('data', 'nikola.novakovic9@gmail.com'): { <same as above> } ...
+
+        self._recent_match_by_profile_pair_and_topic = recent_match_by_profile_pair_and_topic
+        # example: {
+        #   ('data', 'matt.cronin@gmail.com', 'nikolas.novovic9@gmail.com'): { <same as above> } ...
+
+        self._match_requests_profile_to_topic = match_requests_profile_to_topic
+        # example: {
+        #   'sydneynoh@gmail.com': {'javascript', 'html-css', 'python', 'devops'}, ...
+
+        self._match_requests_topic_to_profile = match_requests_topic_to_profile
+        # example: {
+        #   'devops': {'ant@gmail.com', 'colin@gmail.me', 'sydneynoh@gmail.com'}, ...
+
+        self._possible_matches = possible_matches
+        # example: {
+        #   'sydneynoh@gmail.com': {'ant@gmail.com', 'colin@gmail.me', 'sydneynoh@gmail.com'}, ...
+
+    def _pair_score_and_topic(self, p1, p2):
+        p1p2_key = key(p1, p2)
+        if p1p2_key in self._memos_for_pair_score_and_topic:
+            memo = self._memos_for_pair_score_and_topic[p1p2_key]
+        else:
+            memo = self._memos_for_pair_score_and_topic[p1p2_key] = self._not_memoized_pair_score_and_topic(p1, p2)
+
+        return memo
+
+    def _not_memoized_pair_score_and_topic(self, p1, p2):
+        """returns score and topic for this player to player match"""
+        assert p1 is not p2, "must not match a person with themselves"
+
+        score = 1.0  # base score
+
+        p1_topics = self._match_requests_profile_to_topic[p1]
+        p2_topics = self._match_requests_profile_to_topic[p2]
+        topics = p1_topics.intersection(p2_topics)
+
+        # if this fails then we are being very inefficient about picking pairs to test
+        assert len(topics) > 0, "won't score pair with no common topics"
+
+        # promote matching of people who have never participated before or haven't participated recently
```

Comment:
There absolutely is a way to do this. It's good thinking. But I'm not _yet_ going to get creative with the scoring function. There were already several other things that I wanted to do with it but decided not to because it was too complicated. For now I want to release a base scoring function and add complexity as we learn more about the system we're building.